Converte a classificação multi label em várias classificações binárias ou em classificação multi classe

Classifier Chain

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import hamming_loss
from sklearn.multioutput import ClassifierChain
from sklearn.preprocessing import StandardScaler

Quando é multilabel cada classe fica em uma coluna

In [2]:
musica = pd.read_csv('Musica.csv')
musica.head(10)

amazed-suprised  happy-pleased  relaxing-clam  quiet-still  sad-lonely  \
0                0              1              1            0           0   
1                1              0              0            0           0   
2                0              1              0            0           0   
3                0              0              1            0           0   
4                0              0              0            1           0   
5                0              1              1            0           0   
6                1              1              0            0           0   
7                0              0              0            0           0   
8                1              1              0            0           0   
9                0              0              1            1           1   

   angry-aggresive  Mean_Acc1298_Mean_Mem40_Centroid  \
0                0                          0.132498   
1                1                          0.384281   
2                1                          0.541782   
3                0                          0.174288   
4                0                          0.347436   
5                0                          0.228026   
6                0                          0.290836   
7                1                          0.187613   
8                0                          0.384173   
9                0                          0.159915   

   Mean_Acc1298_Mean_Mem40_Rolloff  Mean_Acc1298_Mean_Mem40_Flux  \
0                         0.077848                      0.229227   
1                         0.355249                      0.167190   
2                         0.356491                      0.152246   
3                         0.243935                      0.254326   
4                         0.155448                      0.100047   
5                         0.109648                      0.291862   
6                         0.165289                      0.081240   
7                         0.081515                      0.304717   
8                         0.399660                      0.297172   
9                         0.028334                      0.030533   

   Mean_Acc1298_Mean_Mem40_MFCC_0  ...  Std_Acc1298_Std_Mem40_MFCC_10  \
0                        0.602629  ...                       0.197026   
1                        0.853089  ...                       0.093526   
2                        0.791142  ...                       0.198082   
3                        0.438987  ...                       0.235793   
4                        0.126026  ...                       0.715683   
5                        0.684646  ...                       0.141451   
6                        0.710225  ...                       0.129027   
7                        0.717048  ...                       0.136965   
8                        0.534706  ...                       0.069663   
9                        0.372589  ...                       0.465477   

   Std_Acc1298_Std_Mem40_MFCC_11  Std_Acc1298_Std_Mem40_MFCC_12  \
0                       0.196244                       0.164323   
1                       0.085649                       0.025101   
2                       0.108067                       0.140574   
3                       0.220195                       0.235834   
4                       0.573359                       0.412368   
5                       0.177711                       0.104971   
6                       0.141215                       0.213387   
7                       0.110777                       0.104363   
8                       0.099462                       0.187613   
9                       0.186844                       0.135857   

   BH_LowPeakAmp  BH_LowPeakBPM  BH_HighPeakAmp  BH_HighPeakBPM    BHSUM1  \
0       0.030017       0.253968        0.008473        0.240602  0.136735   
1       0.182955       0.285714        0.156764        0.270677  0.191377   
2       0.099303

In [3]:
classe = musica.iloc[:,0:6].values          #Escolhemos as colunas que contém as classes, ex: amazed-surprised
previsores = musica.iloc[:,7:78].values     #Deixamos as demais colunas dos previsores, como tem 77 colunas, deixamos 78 como final

A padronização serve para que todas as variáveis possam contribuir com o modelo

In [4]:
scaler = StandardScaler()
previsores = scaler.fit_transform(previsores)

In [5]:
previsores

array([[-0.8981437 ,  0.59645542, -0.03960736, ..., -0.49591226,
        -0.69187266, -0.854754  ],
       [ 0.66080303,  0.00318568,  1.37433704, ..., -0.20667729,
        -0.21862043, -0.38517106],
       [ 0.66778286, -0.13972618,  1.02462206, ..., -0.66329082,
        -0.85521088, -0.77487132],
       ...,
       [-1.10473431, -1.13171275, -1.15410898, ..., -0.67620641,
        -0.29937817, -0.74471328],
       [ 0.36627331,  1.10524422,  0.26285006, ...,  0.39361683,
        -0.47519758,  0.45232124],
       [-0.46329804, -0.55724418, -0.19985176, ..., -0.96218602,
        -0.3278917 , -0.78358664]])

In [6]:
X_treinamento, X_teste, y_treinamento, y_teste = train_test_split(previsores, classe, test_size=0.3, random_state=42)     #divisão de dados de treino e teste

In [8]:
svm = SVC()
classifier_chain = ClassifierChain(base_estimator=svm, random_state=42)
classifier_chain.fit(X_treinamento, y_treinamento)

ClassifierChain(base_estimator=SVC(), random_state=42)

In [9]:
y_pred = classifier_chain.predict(X_teste)
loss = hamming_loss(y_teste, y_pred)          #Comparo o que aconteceu com o que foi previsto

print(f'Hamming Loss: {loss}')

Hamming Loss: 0.1900749063670412


O resultado diz que 19% das previsões foram classificadas erroneamente